internal/xds/xdsdependencymanager/xds_dependency_manager.go

```diff
@@ -0,0 +1,241 @@
+	watcher ConfigWatcher
+	// All methods on the xdsResolver type except for the ones invoked by gRPC,
+	// i.e ResolveNow() and Close(), are guaranteed to execute in the context of
+	// this serializer's callback. And since the serializer guarantees mutual
+	// exclusion among these callbacks, we can get by without any mutexes to
+	// access all of the below defined state. The only exception is Close(),
+	// which does access some of this shared state, but it does so after
+	// cancelling the context passed to the serializer.
+	serializer       *grpcsync.CallbackSerializer
+	serializerCancel context.CancelFunc
+
+	// dataplaneAuthority is the authority used for the data plane connections,
+	// which is also used to select the VirtualHost within the xDS
+	// RouteConfiguration.  This is %-encoded to match with VirtualHost Domain
+	// in xDS RouteConfiguration.
+	dataplaneAuthority string
+
+	ldsResourceName       string
+	listenerWatcher       *listenerWatcher
+	currentListenerUpdate *xdsresource.ListenerUpdate
+
+	rdsResourceName    string
+	currentRouteConfig xdsresource.RouteConfigUpdate
+	routeConfigWatcher *routeConfigWatcher
+	currentVirtualHost *xdsresource.VirtualHost
+
+	logger *grpclog.PrefixLogger
+}
+
+// New creates a new dependency manager that manages all the xds resources
+// including LDS, RDS, CDS and EDS and sends update once we have all the
+// resources and sends an error when we get error in listener or route
+// resources.
+func New(listenername, dataplaneAuthority string, xdsClient xdsclient.XDSClient, watcher ConfigWatcher) *DependencyManager {
+	// Builds the dependency manager and starts the listener watch.
+	dm := &DependencyManager{
+		ldsResourceName:    listenername,
+		dataplaneAuthority: dataplaneAuthority,
+		xdsClient:          xdsClient,
+		watcher:            watcher,
+	}
+	dm.logger = prefixLogger(dm)
+
+	// Initialize the serializer used to synchronize the updates from the xDS
+	// client.
+	ctx, cancel := context.WithCancel(context.Background())
+	dm.serializer = grpcsync.NewCallbackSerializer(ctx)
+	dm.serializerCancel = cancel
+
+	// Start the listener watch. Listener watch will start the other resource
+	// watches as needed.
+	dm.listenerWatcher = newListenerWatcher(listenername, dm)
+	return dm
+}
+
+// Close closes the dependency manager and stops all the watchers registered.
+func (m *DependencyManager) Close() {
+	// Cancel the context passed to the serializer and wait for any scheduled
+	// callbacks to complete. Canceling the context ensures that no new
+	// callbacks will be scheduled.
+	m.serializerCancel()
+	<-m.serializer.Done()
+
+	if m.listenerWatcher != nil {
+		m.listenerWatcher.stop()
+	}
+	if m.routeConfigWatcher != nil {
+		m.routeConfigWatcher.stop()
+	}
+}
+
+// ConfigWatcher is notified of the XDSConfig resource updates and errors that
+// are received by the xDS client from the management server. It only receives a
+// XDSConfig update after all the xds resources have been received.
```

Comment:
How about:

// ConfigWatcher is the interface for consumers of aggregated xDS configuration
// from the DependencyManager. The only consumer of this configuration is
// currently the xDS resolver.